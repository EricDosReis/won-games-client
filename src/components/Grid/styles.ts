import styled, { css } from 'styled-components';
import media from 'styled-media-query';

import { GridProps } from '.';

const modifiers = {
  columns: (columns: number) => css`
    grid-template-columns: repeat(${columns}, 1fr);
  `,

  custom: (custom: string) => css`
    grid-template-columns: ${custom};
  `,

  xs: (columns: number) => css`
    ${media.lessThan('small')`
       grid-template-columns: repeat(${columns}, 1fr);
    `}
  `,

  sm: (columns: number) => css`
    ${media.lessThan('medium')`
       grid-template-columns: repeat(${columns}, 1fr);
    `}
  `,

  md: (columns: number) => css`
    ${media.lessThan('large')`
      grid-template-columns: repeat(${columns}, 1fr);
    `}
  `,

  lg: (columns: number) => css`
    ${media.greaterThan('large')`
       grid-template-columns: repeat(${columns}, 1fr);
    `}
  `,
};

export const Wrapper = styled.div<GridProps>`
  ${({ theme, columns, xs, sm, md, lg, custom }) => css`
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(10rem, 1fr));
    gap: ${theme.grid.gutter};

    ${!!columns && modifiers.columns(columns)}
    ${!!custom && modifiers.custom(custom)};
    ${!!xs && modifiers.xs(xs)};
    ${!!sm && modifiers.sm(sm)};
    ${!!md && modifiers.md(md)};
    ${!!lg && modifiers.lg(lg)};
  `};
`;
