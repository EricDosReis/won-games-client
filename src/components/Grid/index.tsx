import React from 'react';

import * as S from './styles';

export interface GridProps {
  columns?: number;
  xs?: number;
  sm?: number;
  md?: number;
  lg?: number;
  custom?: string;
  children: React.ReactNode;
}

const Grid = ({ columns, xs, sm, md, lg, custom, children }: GridProps) => (
  <S.Wrapper columns={columns} custom={custom} xs={xs} sm={sm} md={md} lg={lg}>
    {children}
  </S.Wrapper>
);

export default Grid;
