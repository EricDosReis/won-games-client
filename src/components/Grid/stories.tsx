import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import styled, { css } from 'styled-components';

import Grid, { GridProps } from '.';

export default {
  title: 'Grid',
  component: Grid,
} as Meta;

const Column = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.primary};
    color: ${theme.colors.white};
    padding: ${theme.spacings.xsmall};
    text-align: center;
  `}
`;

export const Basic: Story<GridProps> = args => (
  <Grid {...args}>
    <Column>Column 1</Column>
    <Column>Column 2</Column>
    <Column>Column 3</Column>
    <Column>Column 4</Column>
  </Grid>
);

export const Custom: Story<GridProps> = args => (
  <Grid {...args}>
    <Column>Column 1</Column>
    <Column>Column 2</Column>
    <Column>Column 3</Column>
    <Column>Column 4</Column>
  </Grid>
);

Custom.args = {
  custom: '1fr 20rem 2fr',
};
