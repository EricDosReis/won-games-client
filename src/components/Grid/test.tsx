import { renderWithTheme } from 'utils/tests/helpers';
import Grid from '.';

describe('Component: Grid', () => {
  it('should render Grid', () => {
    const { container } = renderWithTheme(
      <Grid>
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
      </Grid>,
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyle({
      'grid-template-columns': 'repeat(auto-fit,minmax(10rem,1fr))',
      gap: '3.2rem',
    });
  });

  it('should render Grid with columns passed', () => {
    const { container } = renderWithTheme(
      <Grid columns={2}>
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
      </Grid>,
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveStyle({
      'grid-template-columns': 'repeat(2,1fr)',
    });
  });

  it('should render Grid with 1 columns if extra small', () => {
    const { container } = renderWithTheme(
      <Grid xs={1}>
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
      </Grid>,
    );

    expect(container.firstChild).toHaveStyleRule(
      'grid-template-columns',
      'repeat(1,1fr)',
      {
        media: '(max-width:450px)',
      },
    );
  });

  it('should render Grid with 2 columns if small', () => {
    const { container } = renderWithTheme(
      <Grid sm={2}>
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
      </Grid>,
    );

    expect(container.firstChild).toHaveStyleRule(
      'grid-template-columns',
      'repeat(2,1fr)',
      {
        media: '(max-width:768px)',
      },
    );
  });

  it('should render Grid with 3 columns if medium', () => {
    const { container } = renderWithTheme(
      <Grid md={3}>
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
      </Grid>,
    );

    expect(container.firstChild).toHaveStyleRule(
      'grid-template-columns',
      'repeat(3,1fr)',
      {
        media: '(max-width:1170px)',
      },
    );
  });

  it('should render Grid with 4 columns if large', () => {
    const { container } = renderWithTheme(
      <Grid lg={4}>
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
        <p>Content 4</p>
      </Grid>,
    );

    expect(container.firstChild).toHaveStyleRule(
      'grid-template-columns',
      'repeat(4,1fr)',
      {
        media: '(min-width:1170px)',
      },
    );
  });

  it('should render Grid with custom columns setup', () => {
    const { container } = renderWithTheme(
      <Grid custom="1fr 20rem 2fr">
        <p>Content 1</p>
        <p>Content 2</p>
        <p>Content 3</p>
      </Grid>,
    );

    expect(container.firstChild).toHaveStyle({
      'grid-template-columns': '1fr 20rem 2fr',
    });
  });
});
