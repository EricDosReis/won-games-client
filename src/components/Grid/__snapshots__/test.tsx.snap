// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Grid should render Grid 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(auto-fit,minmax(10rem,1fr));
  gap: 3.2rem;
}

<div
  class="c0"
>
  <p>
    Content 1
  </p>
  <p>
    Content 2
  </p>
  <p>
    Content 3
  </p>
</div>
`;

exports[`Component: Grid should render Grid with columns passed 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(auto-fit,minmax(10rem,1fr));
  gap: 3.2rem;
  grid-template-columns: repeat(2,1fr);
}

<div
  class="c0"
>
  <p>
    Content 1
  </p>
  <p>
    Content 2
  </p>
  <p>
    Content 3
  </p>
</div>
`;
